clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1049,15 +1049,24 @@ public void assign(Collection<TopicPartition> partitions) {
             if (partitions == null) {
                 throw new IllegalArgumentException("Topic partition collection to assign to cannot be null");
             } else if (partitions.isEmpty()) {
+                fetcher.clearBufferedData();
                 this.unsubscribe();
             } else {
                 Set<String> topics = new HashSet<>();
+                Set<TopicPartition> topicPartitionsToClearBufferFor = new HashSet<>();
```

Comment:
Could you just pass the new subscription to `clearBufferedDataForTopicPartitions` and clear anything not in the set? That would also handle the other case since we could pass an empty set.

Reply:
Also, do we want similar logic for `subscribe`? We can drop completed fetches if they are for partitions from topics no longer subscribed.

Reply:
I think we can keep the subscribe as it is. The reason I say this is because in case of Mirror-makers, the consumer can undergo multiple rebalances. With that, lets say the consumer was assigned a TopicPartition T0-P0, T0-P1 and it has cached data for both of these.  On first rebalance lets say it gets assigned T0-P1, then we would have cleared the data for T0-P0. At this point lets says it undergoes rebalance and it gets assigned T0-P0, then the earlier cleanup for T0-P0 was unnecessary. Thoughts ??

Reply:
I'm confused about the problem you are trying to solve. The consumer doesn't use this `assign` API itself after completing a rebalance. This API is only used by applications which are using manual partition assignment.

Reply:
Right. I am exactly trying to solve that problem where if applications change the assignment before they have drained out the data for there earlier assignment, we don't need to buffer it and waste time over looping over the unnecessary completedFetches.

Reply:
Yes, then can you explain why `subscribe` does not have a similar concern? If a user changes a subscription and some topics are no longer needed, won't we also waste time looping over unnecessary completed fetches?

Reply:
Yes, same applies to subscribe as well. I think subscribe is used with Mirror-Maker as well I believe and I was thinking, if doing this in subscribe would cause performance regression, because during multiple rebalances in mirror-maker, you might end up cleaning the topic-partition that the consumer host gets ownership for, in the final rebalance.

I don't have a strong opinion on this, we can do the clean up in subscribe as well.

Reply:
@hachikuji  wanted to check if you want me to add this to subscribe as well ? I can upload a new patch for that. Let me know.

Reply:
Yes I agree with Jason that we should apply the same logic to subscribe as well.

Reply:
Cool. Thanks @lindong28 , I will update the patch.